src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,503 @@
+        equations to `Ax = b`. `A` on the left-hand side is a parse matrix representing
+        precomputed coefficients of shift equations; and on the right-side, `b` is
+        estimated 1D shifts along the theta direction between two Fourier rays (one in
+        image i and the other in image j).  Each row of shift equations contains four
+        unknowns, shifts in x, y for a pair of images. The detailed implementation
+        can be found in the book chapter as below:
+        Y. Shkolnisky and A. Singer,
+        Center of Mass Operators for CryoEM - Theory and Implementation,
+        Modeling Nanoscale Imaging in Electron Microscopy,
+        T. Vogt, W. Dahmen, and P. Binev (Eds.)
+        Nanostructure Science and Technology Series,
+        Springer, 2012, pp. 147–177
+
+        :param equations_factor: The factor to rescale the number of shift equations
+            (=1 in default)
+        :param max_memory: If there are N images and N_check selected to check
+            for common lines, then the exact system of equations solved for the shifts
+            is of size 2N x N(N_check-1)/2 (2N unknowns and N(N_check-1)/2 equations).
+            This may be too big if N is large. If `max_memory` is not None (in megabytes),
+            then the algorithm will use `equations_factor` times the total number of
+            equations if the resulting total number of memory requirements is less
+            than `max_memory`; otherwise it will reduce the number of equations to
+            fit in `max_memory`. If `max_memory` is `None`, the shift equations
+            will be generated exactly from the common line matrix instead of estimated
+            rotation matrices.
+        """
+
+        # Generate approximated shift equations from estimated rotations
+        shift_equations, shift_b = self._get_shift_equations_approx(equations_factor, max_memory)
+
+        est_shifts = sparse.linalg.lsqr(shift_equations, shift_b)[0]
+        est_shifts = est_shifts.reshape((2, self.n_img), order='F')
+
+        return est_shifts
+
+    def _get_shift_equations_approx(self, equations_factor=1, max_memory=10000):
+        """
+        Generate approximated shift equations from estimated rotations
+
+        The function computes the common lines from the estimated rotations,
+        and then, for each common line, estimates the 1D shift between its two
+        Fourier rays (one in image i and the other in image j). Using the common
+        lines and the 1D shifts, shift equations are generated randomly based
+        on a memory factor and represented by sparse matrix.
+
+        This function processes the (Fourier transformed) images exactly as the
+        `build_clmatrix` function.
+
+        :param equations_factor: The factor to rescale the number of shift equations
+            (=1 in default)
+        :param max_memory: If there are N images and N_check selected to check
```

Comment:
This is no way to avoid it if n_img is really large. For our current tests we use all the equations to make sure the result can be reproduced. But I can guess the number of equations needed can be different for different resolutions and number of images. In this algorithm level, the researchers should know better.

Reply:
Okay, not entirely sure what you meant. I was just talking about these components which make specific claims regarding memory footprint. It is as you suggest, hard to grok the global program at this time, in part because similar claims cannot always be trusted.

Memory bounding appears to be the main reason this code exists; it might be nice to validate how closely the implementation follows expectation.  Perhaps we can do that as a follow up exercise, I have an idea in mind. The short answer appears to be, no, this was not validated, which is all I needed to know at this time.  Thanks.